staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -913,38 +946,42 @@ func (c *cacheWatcher) stop() {
 	}
 }
 
-func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer, budget *timeBudget) {
-	// Try to send the event immediately, without blocking.
+func (c *cacheWatcher) nonblockingAdd(event *watchCacheEvent) bool {
 	select {
 	case c.input <- event:
-		return
+		return true
 	default:
+		return false
 	}
+}
 
-	// OK, block sending, but only for up to <timeout>.
-	// cacheWatcher.add is called very often, so arrange
-	// to reuse timers instead of constantly allocating.
-	startTime := time.Now()
-	timeout := budget.takeAvailable()
-
-	timer.Reset(timeout)
+func (c *cacheWatcher) add(event *watchCacheEvent, timeLeft *time.Timer) bool {
```

Comment:
Please add a comment saying that nil timer means that add will not block (if it can't send even immediately, it will break the watcher) - this is pretty important.

Reply:
done